sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,766 @@
+# Specialization constants
+
+Specialization constants are implemented in accordance with how they are defined
+by SYCL 2020 specification: [SYCL registry][sycl-registry],
+[direct link to the specification][sycl-2020-spec].
+
+[sycl-registry]: https://www.khronos.org/registry/SYCL/
+[sycl-2020-spec]: https://www.khronos.org/registry/SYCL/specs/sycl-2020/pdf/sycl-2020.pdf
+
+TODO: feature overview? code example?
+
+## Design
+
+[SYCL 2020][sycl-2020-spec] defines specialization constant as:
+
+> A constant variable where the value is not known until compilation of the
+> SYCL kernel function.
+>
+> Glossary
+
+Therefore, implementation is based on [SPIR-V speficiation][spirv-spec] support
+for [Specialization][spirv-specialization].
+
+[spirv-spec]: https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html
+[spirv-specialization]: https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#SpecializationSection
+
+However, the specification also states the following:
+
+> It is expected that many implementations will use an intermediate language
+> representation ... such as SPIR-V, and the intermediate language will have
+> native support for specialization constants. However, implementations that do
+> not have such native support must still support specialization constants in
+> some other way.
+>
+> Section 4.11.12.2. Specialization constant support
+
+Having that said, the following should be implemented:
+
+1. We need to ensure that in generated SPIR-V, calls to
+`get_specialization_constant` are replaced with corresponding instructions for
+referencing specialization constants.
+
+2. SYCL provides a mechanism to specify default values of specialization
+constants, which should be reflected in generated SPIR-V. This part is
+especially tricky, because this happens in host part of the SYCL program, which
+means that without special handling it won't even be visible to device compiler.
+
+3. We need to ensure that DPC++ RT properly set specialization constants used in
+the program: SYCL uses non-type template parameters to identify specialization
+constants in the program, while at SPIR-V and OpenCL level, each specialization
+constant is defined by its numerical ID, which means that we need to maintain
+some mapping from SYCL identifiers to a numeric identifiers to be able to set
+specialization constats. Moreover, at SPIR-V level composite specialization
+constants do not have separate ID and can only be set by setting value to each
+member of a composite, which means that we have 1:n mapping between SYCL
+identifiers and numeric IDs of specialization constants.
+
+4. When AOT compilation is used or target is a CUDA device (where NVPTX
```

Comment:
Why does this impact AOT compilation if specialization constants must be defined by compilation time?

Reply:
The main benefit of spec constants comes from setting their value at runtime, then JITting with spec constant replaced by a the constants, which appears as such to the JITter.

In AOT mode there is no runtime SPIRV JITting, so the above is not possible, hence some other solution is needed (e.g. passed spec constant values via a kernel argument).